proposals/unsafe-evolution.md

```diff
@@ -0,0 +1,138 @@
+types cannot be inherently unsafe. It is only the action of using a type that could be unsafe, not the existence of that type.
+
+In order for this information to flow through the system, we therefore need to have a way to mark methods themselves as `unsafe`. Today, `unsafe` as a method modifier has no external impact,
+it only allows pointers to be used in the body of the member. Going forward, `unsafe` as a modifier will actually publicly change the meaning of the member; it will indicate that the member
+has memory safety concerns and any usages must be manually validated by the programmer using the member.
+
+This is a potentially large breaking change for particular segments of the C# user base. Our hope is that, for many of our users, this is effectively transparent, and updating to the new rules
+will be seemless. However, given that some large API surfaces like large parts of reflection may need to be marked `unsafe`, we do think it likely that there will need to be a decent on-ramp to
+the new rules to avoid entirely bifurcating the ecosystem.
+
+## Detailed Design
+
+### Existing `unsafe` rules
+
+The existing C# specification has a large section devoted to `unsafe`: [§24 Unsafe code][unsafe-code.md]. It is defined as conditionally normative, as it is not required for a valid C# compiler
+to support the `unsafe` feature. Much of what is currently considered conditionally normative will no longer be so after this change, as most of the definition of pointers is no longer considered
+unsafe in itself. [Pointer types][pointer-types-spec], [Fixed and moveable variables][fixed-and-moveable-variables], all [pointer expressions][pointer-expressions] (except for
+[pointer indirection][pointer-indirection], [pointer member access][pointer-member-access], and [pointer element access][pointer-element-access]), and [the `fixed` statement][fixed-statement]
+are all no longer considered `unsafe`, and exist in normal C# with no requirement to be used in an `unsafe` context. Similarly, declaring a [fixed size buffer][fixed-size-buffer-declarations] or
+an initialized [`stackalloc`][stack-allocation-spec] are also perfectly legal in safe C#. For all of these cases, it is only _accessing_ the memory that is unsafe.
+
+Given the extensive rewrite of both the `unsafe` code section and other parts C# specification inherent in this change, it would be unwieldy and likely not useful to provide a line-by-line diff
+of the existing rules of the specification. Instead, we will provide an overview of the change to make in a given section, as well as specific new rules for what is allowed in `unsafe` contexts.
+
+#### Expression safety state
+
+> [!NOTE]
+> Now accepting naming suggestions
+
+We introduce a new state that is tracked for all expressions in C#: the safety state. There are two possible safety states for any expression: safe, or unsafe. Expressions with a safety state of
```

Comment:
> the safety state

Perhaps "memory-safety" state to better distinguish it from the "ref-safety" term.